Yarn 2/Yarn 3 documentation page, in order to update node buildpack to get this compatibility (#1482)
Scalingo/documentation: _posts/languages/nodejs/2000-01-01-yarn-2.md

---
title: Node.js
nav: Yarn 2 and Yarn 3 Compatibility
modified_at: 2021-12-10 00:00:00
tags: nodejs
---

# What's new in new Yarn major version

> Yarn was developed as an alternative way to handle Node.JS dependencies to NPM

Reviewer: ```suggestion
Yarn was developed as an alternative way to handle Node.js dependencies to NPM
```

but was fully compatible with it. They were both manipulating the
`node_modules` directory and switching to one or the other tool was
straightforward.

Yarn 2 and Yarn 3 have evolved to use another approached, aiming at making

> dependency management for Node.JS application more efficient, especially during

Reviewer: ```suggestion
dependency management for Node.js application more efficient, especially during
```

its deployment.

{% warning %}

> This article only covers application using the Node.JS buildpack and not other
> buildpacks which include installation of a Node.JS stack like Ruby or PHP. If
> Yarn 2 or 3 is used alongside a Ruby or Node.JS application, the
> [multi-buildpacks]({% post_url platform/deployment/buildpacks/2000-01-01-multi
> %}) should be used.

Reviewer: ```suggestion
This article only covers application using the Node.js buildpack and not other
buildpacks which include installation of a Node.js stack like Ruby or PHP. If
Yarn 2 or 3 is used alongside a Ruby or PHP application, the
[multi-buildpacks]({% post_url platform/deployment/buildpacks/2000-01-01-multi
%}) should be used.
```

And also, maybe target this doc page for PHP apps: https://doc.scalingo.com/languages/php/php-nodejs?

{% endwarning %}

> You are not force to migrated to Yarn 2 or 3, compatibility with Yarn 1 will be

Reviewer: ```suggestion
You are not forced to migrate to Yarn 2 or 3, compatibility with Yarn 1 will be
```

> kept as long as possible, but to enjoy fastest deployments and last patches,

Reviewer: ```suggestion
kept as long as possible. But to enjoy fastest deployments and last patches,
```

you are encouraged to upgrade your project or get back to NPM.

New versions of Yarn have 2 ways to work:

* With Plug'n Play: it is expected to checkout the application dependencies
  `.yarn/cache` in your git repository. It will accelerate deployment since it

>   won't be necessary to download dependencies they are already in your repository.

Reviewer: ```suggestion
  won't be necessary to download dependencies that are already in your repository.
```

>   Don't worry yarn is optimizing and compressing them to prevent using too much space.

Reviewer: Just for consistency as you spelled it that way in the page:
```suggestion
  Don't worry Yarn is optimizing and compressing them to prevent using too much space.
```

  No `node_modules` directory will be present at the end of the build.

* Without Plug'n Play: Dependencies will be downloaded normally during the
  deployment and linked into the `node_modules` directory, it's
  retrocompatible with most tools you have been using.

> More details in the [Yarn official documentation](https://yarnpkg.com/getting-started/migration)

Reviewer: ```suggestion
More details in the [Yarn official documentation](https://yarnpkg.com/getting-started/migration).
```

# Setting up your project

## 1. Switch to the last stable Yarn version

> Configure Yarn to user it's more recent ≥ v2.0.0 release

Reviewer: ```suggestion
Configure Yarn to use its more recent ≥ v2.0.0 release
```

```
yarn set version berry
yarn install
```

The file `yarn.lock` is updated and new files are added to the `.yarn` directory.

## 2a. Configure with Plug'n Play enabeld

### Update your `.gitignore` file

> Modify your `.gitignore` to ensure only required files are commited into

Reviewer: ```suggestion
Modify your `.gitignore` to ensure only required files are committed into
```

your repository, add the following:

```
.yarn/*
!.yarn/cache
!.yarn/patches
!.yarn/plugins
!.yarn/releases
!.yarn/sdks

> !.yarn/versions

Reviewer: I didn't know this syntax in gitignore, nice!

```

### Disable dependency caching in your app

```
scalingo --app my-app env-set NODE_MODULES_CACHE=false
```

### Add checkout config, metadata and cache files

```
git add yarn.lock .yarn .pnp.*
git commit -m "Deploy with Yarn 3 with Plug'n Play mode"
```

## 2b. OR - Configure without Plug'n Play enabled (retrocompatibily mode)

### Update your `.gitignore` file

> Modify your `.gitignore` to ensure only required files are commited into

Reviewer: ```suggestion
Modify your `.gitignore` to ensure only required files are committed into
```
